File: src/main/java/org/moinex/ui/dialog/investment/BondInterestHistoryController.java
/*
 * Filename: BondInterestHistoryController.java
 * Created on: February 21, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.investment;

import java.math.BigDecimal;
import java.text.MessageFormat;
import java.util.List;
import javafx.beans.property.SimpleStringProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.fxml.FXML;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.control.TextInputDialog;
import lombok.NoArgsConstructor;
import org.moinex.model.investment.Bond;
import org.moinex.model.investment.BondInterestCalculation;
import org.moinex.service.BondInterestCalculationService;
import org.moinex.service.I18nService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

@Controller
@NoArgsConstructor
public class BondInterestHistoryController {

    @FXML private Label bondNameLabel;
    @FXML private Label bondTypeLabel;
    @FXML private Label bondInterestTypeLabel;
    @FXML private TableView<BondInterestCalculation> historyTable;
    @FXML private Button editButton;
    @FXML private Button resetButton;
    @FXML private Button recalculateButton;
    @FXML private Button closeButton;

    private Bond bond;
    private BondInterestCalculationService bondInterestCalculationService;
    private I18nService i18nService;

    private static final Logger logger =
            LoggerFactory.getLogger(BondInterestHistoryController.class);

    @Autowired
    public BondInterestHistoryController(
            BondInterestCalculationService bondInterestCalculationService,
            I18nService i18nService) {
        this.bondInterestCalculationService = bondInterestCalculationService;
        this.i18nService = i18nService;
    }

    public void setBond(Bond bond) {
        this.bond = bond;
        initialize();
    }

    @FXML
    private void initialize() {
        if (bond == null) {
            return;
        }

        setupBondInfo();
        setupTableColumns();
        loadHistoryData();
        setupTableListeners();
    }

    private void setupBondInfo() {
        bondNameLabel.setText(bond.getName());
        bondTypeLabel.setText(bond.getType().toString());
        bondInterestTypeLabel.setText(bond.getInterestType() + " - " + bond.getInterestIndex());
    }

    private void setupTableColumns() {
        historyTable.getColumns().clear();

        TableColumn<BondInterestCalculation, String> monthColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_INTEREST_HISTORY_TABLE_MONTH));
        monthColumn.setCellValueFactory(
                cellData ->
                        new SimpleStringProperty(
                                cellData.getValue().getReferenceMonth().toString()));

        TableColumn<BondInterestCalculation, String> quantityColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_INTEREST_HISTORY_TABLE_QUANTITY));
        quantityColumn.setCellValueFactory(
                cellData -> {
                    BigDecimal quantity = cellData.getValue().getQuantity();
                    return new SimpleStringProperty(UIUtils.formatCurrency(quantity));
                });

        TableColumn<BondInterestCalculation, String> investedAmountColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .BOND_INTEREST_HISTORY_TABLE_INVESTED_AMOUNT));
        investedAmountColumn.setCellValueFactory(
                cellData -> {
                    BigDecimal investedAmount = cellData.getValue().getInvestedAmount();
                    return new SimpleStringProperty(UIUtils.formatCurrency(investedAmount));
                });

        TableColumn<BondInterestCalculation, String> monthlyInterestColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .BOND_INTEREST_HISTORY_TABLE_MONTHLY_INTEREST));
        monthlyInterestColumn.setCellValueFactory(
                cellData -> {
                    BigDecimal monthlyInterest = cellData.getValue().getMonthlyInterest();
                    return new SimpleStringProperty(UIUtils.formatCurrency(monthlyInterest));
                });

        TableColumn<BondInterestCalculation, String> accumulatedInterestColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .BOND_INTEREST_HISTORY_TABLE_ACCUMULATED_INTEREST));
        accumulatedInterestColumn.setCellValueFactory(
                cellData -> {
                    BigDecimal accumulatedInterest = cellData.getValue().getAccumulatedInterest();
                    return new SimpleStringProperty(UIUtils.formatCurrency(accumulatedInterest));
                });

        TableColumn<BondInterestCalculation, String> finalValueColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_INTEREST_HISTORY_TABLE_FINAL_VALUE));
        finalValueColumn.setCellValueFactory(
                cellData -> {
                    BigDecimal finalValue = cellData.getValue().getFinalValue();
                    return new SimpleStringProperty(UIUtils.formatCurrency(finalValue));
                });

        TableColumn<BondInterestCalculation, String> statusColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_INTEREST_HISTORY_TABLE_STATUS));
        statusColumn.setCellValueFactory(
                cellData -> {
                    if (cellData.getValue().isManuallyAdjusted()) {
                        return new SimpleStringProperty(
                                i18nService.tr(
                                        Constants.TranslationKeys
                                                .BOND_INTEREST_HISTORY_STATUS_ADJUSTED));
                    }
                    return new SimpleStringProperty(
                            i18nService.tr(
                                    Constants.TranslationKeys
                                            .BOND_INTEREST_HISTORY_STATUS_AUTOMATIC));
                });

        historyTable
                .getColumns()
                .addAll(
                        monthColumn,
                        quantityColumn,
                        investedAmountColumn,
                        monthlyInterestColumn,
                        accumulatedInterestColumn,
                        finalValueColumn,
                        statusColumn);
    }

    private void loadHistoryData() {
        List<BondInterestCalculation> history =
                bondInterestCalculationService.getMonthlyInterestHistory(bond);
        ObservableList<BondInterestCalculation> observableList =
                FXCollections.observableArrayList(history);
        historyTable.setItems(observableList);
    }

    private void setupTableListeners() {
        historyTable
                .getSelectionModel()
                .selectedItemProperty()
                .addListener(
                        (obs, oldVal, newVal) -> {
                            boolean hasSelection = newVal != null;
                            editButton.setDisable(!hasSelection);
                            resetButton.setDisable(!hasSelection || !newVal.isManuallyAdjusted());
                        });

        editButton.setDisable(true);
        resetButton.setDisable(true);
    }

    @FXML
    private void handleEditInterest() {
        BondInterestCalculation selected = historyTable.getSelectionModel().getSelectedItem();
        if (selected == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_NO_MONTH_SELECTED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_NO_MONTH_SELECTED_MESSAGE));
            return;
        }

        TextInputDialog dialog = new TextInputDialog(selected.getMonthlyInterest().toString());
        dialog.setTitle(
                i18nService.tr(
                        Constants.TranslationKeys
                                .BOND_INTEREST_HISTORY_DIALOG_EDIT_INTEREST_TITLE));
        dialog.setHeaderText(
                MessageFormat.format(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .BOND_INTEREST_HISTORY_DIALOG_EDIT_INTEREST_HEADER),
                        selected.getReferenceMonth().toString()));
        dialog.setContentText(
                i18nService.tr(
                        Constants.TranslationKeys
                                .BOND_INTEREST_HISTORY_DIALOG_EDIT_INTEREST_CONTENT));

        var result = dialog.showAndWait();
        if (result.isEmpty()) {
            return;
        }

        try {
            BigDecimal newValue = new BigDecimal(result.get());

            if (newValue.compareTo(BigDecimal.ZERO) < 0) {
                WindowUtils.showErrorDialog(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .BOND_INTEREST_HISTORY_DIALOG_INVALID_VALUE_TITLE),
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .BOND_INTEREST_HISTORY_DIALOG_INVALID_VALUE_MESSAGE));
                return;
            }

            bondInterestCalculationService.adjustMonthlyInterest(
                    bond.getId(), selected.getReferenceMonth(), newValue);

            WindowUtils.showSuccessDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_INTEREST_ADJUSTED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_INTEREST_ADJUSTED_MESSAGE));

            loadHistoryData();
            logger.info(
                    "Interest adjusted for bond {} month {}: {}",
                    bond.getName(),
                    selected.getReferenceMonth(),
                    newValue);

        } catch (NumberFormatException e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_INVALID_NUMBER_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_INVALID_NUMBER_MESSAGE));
            logger.warn("Invalid number format for interest adjustment", e);
        } catch (Exception e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_ERROR_ADJUSTING_TITLE),
                    e.getMessage());
            logger.error("Error adjusting interest", e);
        }
    }

    @FXML
    private void handleResetToAutomatic() {
        BondInterestCalculation selected = historyTable.getSelectionModel().getSelectedItem();
        if (selected == null || !selected.isManuallyAdjusted()) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_NO_ADJUSTED_MONTH_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_NO_ADJUSTED_MONTH_MESSAGE));
            return;
        }

        if (!WindowUtils.showConfirmationDialog(
                i18nService.tr(
                        Constants.TranslationKeys
                                .BOND_INTEREST_HISTORY_DIALOG_CONFIRM_RESET_HEADER),
                MessageFormat.format(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .BOND_INTEREST_HISTORY_DIALOG_CONFIRM_RESET_MESSAGE),
                        selected.getReferenceMonth().toString()))) {
            return;
        }

        try {
            bondInterestCalculationService.resetToAutomaticCalculation(
                    bond.getId(), selected.getReferenceMonth());

            WindowUtils.showSuccessDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_INTEREST_RESET_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_INTEREST_RESET_MESSAGE));

            loadHistoryData();
            logger.info(
                    "Interest reset to automatic for bond {} month {}",
                    bond.getName(),
                    selected.getReferenceMonth());

        } catch (Exception e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_ERROR_RESETTING_TITLE),
                    e.getMessage());
            logger.error("Error resetting interest to automatic", e);
        }
    }

    @FXML
    private void handleRecalculateHistory() {
        if (!WindowUtils.showConfirmationDialog(
                i18nService.tr(
                        Constants.TranslationKeys
                                .BOND_INTEREST_HISTORY_DIALOG_CONFIRM_RECALCULATE_HEADER),
                i18nService.tr(
                        Constants.TranslationKeys
                                .BOND_INTEREST_HISTORY_DIALOG_CONFIRM_RECALCULATE_MESSAGE))) {
            return;
        }

        try {
            bondInterestCalculationService.recalculateAllMonthlyInterest(bond.getId());

            WindowUtils.showSuccessDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_HISTORY_RECALCULATED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_HISTORY_RECALCULATED_MESSAGE));

            loadHistoryData();
            logger.info("Interest history recalculated for bond {}", bond.getName());

        } catch (Exception e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_INTEREST_HISTORY_DIALOG_ERROR_RECALCULATING_TITLE),
                    e.getMessage());
            logger.error("Error recalculating interest history", e);
        }
    }

    @FXML
    private void handleClose() {
        closeButton.getScene().getWindow().hide();
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/investment/view/BondInterestHistoryController.kt
/*
 * Filename: BondInterestHistoryController.kt (original filename: BondInterestHistoryController.java)
 * Created on: February 21, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.investment.view

import javafx.beans.property.SimpleStringProperty
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.control.TableColumn
import javafx.scene.control.TableView
import javafx.scene.control.TextInputDialog
import org.moinex.common.constants.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.investment.Bond
import org.moinex.model.investment.BondInterestCalculation
import org.moinex.service.PreferencesService
import org.moinex.service.investment.BondInterestCalculationService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Controller
import java.math.BigDecimal
import java.text.MessageFormat

@Controller
class BondInterestHistoryController(
    private val bondInterestCalculationService: BondInterestCalculationService,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var bondNameLabel: Label

    @FXML
    private lateinit var bondTypeLabel: Label

    @FXML
    private lateinit var bondInterestTypeLabel: Label

    @FXML
    private lateinit var historyTable: TableView<BondInterestCalculation>

    @FXML
    private lateinit var editButton: Button

    @FXML
    private lateinit var resetButton: Button

    @FXML
    private lateinit var recalculateButton: Button

    @FXML
    private lateinit var closeButton: Button

    private lateinit var bond: Bond

    companion object {
        private val logger = LoggerFactory.getLogger(BondInterestHistoryController::class.java)
    }

    fun setBond(b: Bond) {
        bond = b
        initialize()
    }

    @FXML
    private fun initialize() {
        if (!::bond.isInitialized) {
            return
        }

        setupBondInfo()
        setupTableColumns()
        loadHistoryData()
        setupTableListeners()
    }

    private fun setupBondInfo() {
        bondNameLabel.text = bond.name
        bondTypeLabel.text = bond.type.toString()
        bondInterestTypeLabel.text = "${bond.interestType} - ${bond.interestIndex}"
    }

    private fun setupTableColumns() {
        historyTable.columns.clear()

        val monthColumn =
            TableColumn<BondInterestCalculation, String>(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_TABLE_MONTH),
            ).apply {
                setCellValueFactory { cellData ->
                    SimpleStringProperty(cellData.value.referenceMonth.toString())
                }
            }

        val quantityColumn =
            TableColumn<BondInterestCalculation, String>(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_TABLE_QUANTITY),
            ).apply {
                setCellValueFactory { cellData ->
                    val quantity = cellData.value.quantity
                    SimpleStringProperty(UIUtils.formatCurrency(quantity))
                }
            }

        val investedAmountColumn =
            TableColumn<BondInterestCalculation, String>(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_TABLE_INVESTED_AMOUNT),
            ).apply {
                setCellValueFactory { cellData ->
                    val investedAmount = cellData.value.investedAmount
                    SimpleStringProperty(UIUtils.formatCurrency(investedAmount))
                }
            }

        val monthlyInterestColumn =
            TableColumn<BondInterestCalculation, String>(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_TABLE_MONTHLY_INTEREST),
            ).apply {
                setCellValueFactory { cellData ->
                    val monthlyInterest = cellData.value.monthlyInterest
                    SimpleStringProperty(UIUtils.formatCurrency(monthlyInterest))
                }
            }

        val accumulatedInterestColumn =
            TableColumn<BondInterestCalculation, String>(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_TABLE_ACCUMULATED_INTEREST),
            ).apply {
                setCellValueFactory { cellData ->
                    val accumulatedInterest = cellData.value.accumulatedInterest
                    SimpleStringProperty(UIUtils.formatCurrency(accumulatedInterest))
                }
            }

        val finalValueColumn =
            TableColumn<BondInterestCalculation, String>(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_TABLE_FINAL_VALUE),
            ).apply {
                setCellValueFactory { cellData ->
                    val finalValue = cellData.value.finalValue
                    SimpleStringProperty(UIUtils.formatCurrency(finalValue))
                }
            }

        val statusColumn =
            TableColumn<BondInterestCalculation, String>(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_TABLE_STATUS),
            ).apply {
                setCellValueFactory { cellData ->
                    if (cellData.value.isManuallyAdjusted()) {
                        SimpleStringProperty(
                            preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_STATUS_ADJUSTED),
                        )
                    } else {
                        SimpleStringProperty(
                            preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_STATUS_AUTOMATIC),
                        )
                    }
                }
            }

        historyTable.columns.addAll(
            monthColumn,
            quantityColumn,
            investedAmountColumn,
            monthlyInterestColumn,
            accumulatedInterestColumn,
            finalValueColumn,
            statusColumn,
        )
    }

    private fun loadHistoryData() {
        val history = bondInterestCalculationService.getMonthlyInterestHistory(bond)
        val observableList = FXCollections.observableArrayList(history)
        historyTable.items = observableList
    }

    private fun setupTableListeners() {
        historyTable.selectionModel.selectedItemProperty().addListener { _, _, newVal ->
            val hasSelection = newVal != null
            editButton.isDisable = !hasSelection
            resetButton.isDisable = !hasSelection || !newVal.isManuallyAdjusted()
        }

        editButton.isDisable = true
        resetButton.isDisable = true
    }

    @FXML
    private fun handleEditInterest() {
        val selected = historyTable.selectionModel.selectedItem
        if (selected == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_NO_MONTH_SELECTED_TITLE),
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_NO_MONTH_SELECTED_MESSAGE),
            )
            return
        }

        val dialog = TextInputDialog(selected.monthlyInterest.toString())
        dialog.title = preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_EDIT_INTEREST_TITLE)
        dialog.headerText =
            MessageFormat.format(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_EDIT_INTEREST_HEADER),
                selected.referenceMonth.toString(),
            )
        dialog.contentText =
            preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_EDIT_INTEREST_CONTENT)

        val result = dialog.showAndWait()
        if (result.isEmpty) {
            return
        }

        runCatching {
            val newValue = BigDecimal(result.get())

            if (newValue < BigDecimal.ZERO) {
                WindowUtils.showErrorDialog(
                    preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_INVALID_VALUE_TITLE),
                    preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_INVALID_VALUE_MESSAGE),
                )
                return
            }

            bondInterestCalculationService.adjustMonthlyInterest(
                bond.id!!,
                selected.referenceMonth,
                newValue,
            )

            WindowUtils.showSuccessDialog(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_INTEREST_ADJUSTED_TITLE),
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_INTEREST_ADJUSTED_MESSAGE),
            )

            loadHistoryData()
            logger.info(
                "Interest adjusted for bond {} month {}: {}",
                bond.name,
                selected.referenceMonth,
                newValue,
            )
        }.onFailure { e ->
            when (e) {
                is NumberFormatException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_INVALID_NUMBER_TITLE),
                        preferencesService.translate(
                            TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_INVALID_NUMBER_MESSAGE,
                        ),
                    )
                    logger.warn("Invalid number format for interest adjustment", e)
                }
                else -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(
                            TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_ERROR_ADJUSTING_TITLE,
                        ),
                        e.message ?: "",
                    )
                    logger.error("Error adjusting interest", e)
                }
            }
        }
    }

    @FXML
    private fun handleResetToAutomatic() {
        val selected = historyTable.selectionModel.selectedItem
        if (selected == null || !selected.isManuallyAdjusted()) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_NO_ADJUSTED_MONTH_TITLE),
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_NO_ADJUSTED_MONTH_MESSAGE),
            )
            return
        }

        if (!WindowUtils.showConfirmationDialog(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_CONFIRM_RESET_HEADER),
                MessageFormat.format(
                    preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_CONFIRM_RESET_MESSAGE),
                    selected.referenceMonth.toString(),
                ),
            )
        ) {
            return
        }

        runCatching {
            bondInterestCalculationService.resetToAutomaticCalculation(
                bond.id!!,
                selected.referenceMonth,
            )

            WindowUtils.showSuccessDialog(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_INTEREST_RESET_TITLE),
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_INTEREST_RESET_MESSAGE),
            )

            loadHistoryData()
            logger.info(
                "Interest reset to automatic for bond {} month {}",
                bond.name,
                selected.referenceMonth,
            )
        }.onFailure { e ->
            WindowUtils.showErrorDialog(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_ERROR_RESETTING_TITLE),
                e.message ?: "",
            )
            logger.error("Error resetting interest to automatic", e)
        }
    }

    @FXML
    private fun handleRecalculateHistory() {
        if (!WindowUtils.showConfirmationDialog(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_CONFIRM_RECALCULATE_HEADER),
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_CONFIRM_RECALCULATE_MESSAGE),
            )
        ) {
            return
        }

        runCatching {
            bondInterestCalculationService.recalculateAllMonthlyInterest(bond.id!!)

            WindowUtils.showSuccessDialog(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_HISTORY_RECALCULATED_TITLE),
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_HISTORY_RECALCULATED_MESSAGE),
            )

            loadHistoryData()
            logger.info("Interest history recalculated for bond {}", bond.name)
        }.onFailure { e ->
            WindowUtils.showErrorDialog(
                preferencesService.translate(TranslationKeys.BOND_INTEREST_HISTORY_DIALOG_ERROR_RECALCULATING_TITLE),
                e.message ?: "",
            )
            logger.error("Error recalculating interest history", e)
        }
    }

    @FXML
    private fun handleClose() {
        closeButton.scene.window.hide()
    }
}
